bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/verification/DefaultMismatchDetector.java

```diff
@@ -199,6 +180,106 @@ private void verifyResults(
     }
   }
 
+  @Override
+  public ScannerResultVerifier createScannerResultVerifier(Scan request, int maxBufferedResults) {
+    return new DefaultScannerResultVerifier(request, maxBufferedResults);
+  }
+
+  public class DefaultScannerResultVerifier implements ScannerResultVerifier {
+    private final LinkedList<Result> primaryMismatchBuffer;
+    private final LinkedList<Result> secondaryMismatchBuffer;
+    private final int sizeLimit;
+    private final Scan scanRequest;
+
+    private DefaultScannerResultVerifier(Scan scan, int sizeLimit) {
+      this.scanRequest = scan;
+      this.sizeLimit = sizeLimit;
+      this.primaryMismatchBuffer = new LinkedList<>();
+      this.secondaryMismatchBuffer = new LinkedList<>();
+    }
+
+    @Override
+    public void flush() {
+      this.shrinkBuffer(this.primaryMismatchBuffer, 0);
+      this.shrinkBuffer(this.secondaryMismatchBuffer, 0);
+    }
+
+    @Override
+    public void verify(Result[] primary, Result[] secondary) {
+      this.addNotNull(this.primaryMismatchBuffer, primary);
+      this.addNotNull(this.secondaryMismatchBuffer, secondary);
+
+      this.matchResults();
+
+      this.shrinkBuffer(this.primaryMismatchBuffer, this.sizeLimit);
+      this.shrinkBuffer(this.secondaryMismatchBuffer, this.sizeLimit);
+    }
+
+    private void addNotNull(Deque<Result> mismatchBuffer, Result[] results) {
+      for (Result result : results) {
+        if (result == null) break;
+        mismatchBuffer.addLast(result);
+      }
+    }
+
+    private void matchResults() {
+      Set<Result> matches = this.matchingResults();
+      while (!matches.isEmpty()) {
+        Result primaryResult = this.firstMatch(this.primaryMismatchBuffer, matches);
+        this.popUntilMatchesWanted(this.secondaryMismatchBuffer, primaryResult, matches);
+        metricsRecorder.recordReadMatches(HBaseOperation.NEXT, 1);
+      }
+    }
+
+    private Result firstMatch(Deque<Result> primaryBuffer, Set<Result> matches) {
+      while (!primaryBuffer.isEmpty()) {
+        Result result = primaryBuffer.removeFirst();
+        if (matches.remove(result)) {
+          return result;
+        }
+        logAndRecordScanMismatch(result);
+      }
+      // A match should have been found.
+      assert false;
+      return null;
+    }
+
+    void popUntilMatchesWanted(Deque<Result> buffer, Result wanted, Set<Result> matches) {
+      while (!buffer.isEmpty()) {
+        Result result = buffer.removeFirst();
+        matches.remove(result);
+        if (Comparators.resultsEqual(result, wanted)) {
+          return;
+        }
+        logAndRecordScanMismatch(result);
+      }
+      // Wanted result should have been found.
+      assert false;
+    }
+
+    private Set<Result> matchingResults() {
+      Set<Result> results = new HashSet<>(this.primaryMismatchBuffer);
```

Comment:
I think that this Set can be calculated on the fly instead of rebuilding it every time.